app/controllers/foreman_tasks/api/tasks_controller.rb

```diff
@@ -122,6 +121,42 @@ def bulk_resume
         }
       end
 
+      api :POST, '/tasks/bulk_cancel', N_('Cancel all cancellable tasks')
+      param :search, String, :desc => N_('Cancel tasks matching search string')
+      param :task_ids, Array, :desc => N_('Cancel specific tasks by ID')
+      def bulk_cancel
+        scope = resource_scope
+        scope = scope.search_for(params[:search]) if params[:search]
+        scope = scope.select('DISTINCT foreman_tasks_tasks.*')
+        if params[:search].nil? && params[:task_ids].nil?
+          raise ArgumentError, _('Please provide at least one of search or task_ids parameters in the request')
+        end
+        scope = scope.where(:id => params[:task_ids]) if params[:task_ids]
+
+        cancelled = []
+        failed = []
+        skipped = []
+        scope.each do |task|
+          if task.cancellable?
+            begin
+              task.cancel
+              cancelled << task_hash(task)
+            rescue RuntimeError
+              failed << task_hash(task)
+            end
+          else
+            skipped << task_hash(task)
+          end
+        end
```

Comment:
How about using `Enumerable#partition` to first split the scope into tasks which can be cancelled and which cannot and then operating only on those which can be cancelled?

Reply:
Looks much better that way, thanks!